### PR TITLE
VSTS-211 - Fixed a bug where if SONAR_SCANNER_OPTS is null after the …

### PR DESCRIPTION
**Pull request by mickael-caro-sonarsource**
…retrieval of its initial value, will fail the concat call.

### CODE REVIEW

File: common/ts/prepare-task.ts [hunk 1]
```diff
@@ -54,6 +54,9 @@ export default async function prepareTask(endpoint: Endpoint, rootPath: string)
 
   if (scannerMode === ScannerMode.CLI) {
     let actualSonarOpts = tl.getVariable('SONAR_SCANNER_OPTS');
+    if (actualSonarOpts == null) {
```
Comment (gregaubert) [inline common/ts/prepare-task.ts]: No need to change it, but you can write short things like `let actualSonarOpts = tl.getVariable('SONAR_SCANNER_OPTS') || '';` for the same result

File: package.json [hunk 2]
```diff
@@ -91,6 +91,7 @@
     "singleQuote": true
   },
   "dependencies": {
+    "@types/jest-when": "2.4.1",
```
Comment (gregaubert) [inline package.json]: You probably want this in `devDependencies` instead

File: common/ts/__tests__/publish-task-test.ts [hunk 3]
```diff
@@ -60,8 +60,18 @@ it('check multiple report status and set global quality gate for build propertie
   taskReportArray.push(taskReport);
   taskReportArray.push(taskReport);
 
-  jest.spyOn(Task, 'waitForTaskCompletion').mockImplementation(() => returnedTaskOk);
-  jest.spyOn(TaskReport, 'createTaskReportsFromFiles').mockImplementation(() => taskReportArray);
+  jest.spyOn(Task, 'waitForTaskCompletion').mockImplementation(
+    () =>
+      new Promise<Task>(resolve => {
+        return resolve(returnedTaskOk);
```
Comment (gregaubert) [inline common/ts/__tests__/publish-task-test.ts]: You can do things like `jest.spyOn(Task, 'waitForTaskCompletion').mockResolvedValue(returnedTaskOk);` and it should have the same effect.

File: common/ts/__tests__/publish-task-test.ts [hunk 4]
```diff
@@ -72,14 +82,29 @@ it('check multiple report status and set global quality gate for build propertie
     null
   );
 
-  jest.spyOn(Analysis, 'getAnalysis').mockImplementationOnce(() => returnedAnalysisOk);
-  jest.spyOn(Analysis, 'getAnalysis').mockImplementationOnce(() => returnedAnalysisOk);
+  jest.spyOn(Analysis, 'getAnalysis').mockImplementationOnce(
```
Comment (gregaubert) [inline common/ts/__tests__/publish-task-test.ts]: There is also a `jest.spyOn(Analysis, 'getAnalysis').mockResolvedValueOnce(returnedAnalysisOk);`

File: common/ts/__tests__/prepare-task-test.ts [hunk 5]
```diff
@@ -24,8 +26,18 @@ it('should display warning for dedicated extension for Sonarcloud', async () =>
   jest.spyOn(tl, 'warning').mockImplementation(() => null);
   jest.spyOn(Scanner, 'getPrepareScanner').mockImplementation(() => scannerObject);
   jest.spyOn(scannerObject, 'runPrepare').mockImplementation(() => null);
-  jest.spyOn(request, 'getServerVersion').mockImplementation(() => '7.0.0');
-  jest.spyOn(prept, 'getDefaultBranch').mockImplementation(() => 'refs/heads/master');
+  jest.spyOn(request, 'getServerVersion').mockImplementation(
```
Comment (gregaubert) [inline common/ts/__tests__/prepare-task-test.ts]: Same in this file.

File: common/ts/prepare-task.ts (hunk 1 above)
Comment (O-heu) [inline common/ts/prepare-task.ts]: We get an error message with undefined, does this null check fixes that?

Comment (mickael-caro-sonarsource) [reply common/ts/prepare-task.ts]: I knew the syntax in C# but i was looking for something like this, thanks !

File: package.json (hunk 2 above)
Comment (mickael-caro-sonarsource) [reply package.json]: Ok, i moved all dependencies in dev for jest stuff.

File: common/ts/__tests__/publish-task-test.ts (hunk 3 above)
Comment (mickael-caro-sonarsource) [reply common/ts/__tests__/publish-task-test.ts]: Thanks, that works !

File: common/ts/prepare-task.ts (hunk 1 above)
Comment (mickael-caro-sonarsource) [reply common/ts/prepare-task.ts]: Yes !

**Pull request merged**